makefiles/ubsan.inc.mk

```diff
@@ -0,0 +1,61 @@
+# Copyright (C) 2019 Kaspar Schleiser <kaspar@schleiser.de>
+#
+# # Introduction
+#
+# This file contains makefile convenience support for gcc/clang's undefined
+# behaviour sanitizer.
+#
+# # Overview
+#
+# Both gcc and clang allow generation on code that does runtime checks for
+# undefined behavior (UB).
+#
+# E.g., the following code might trigger UB for some parameters:
+#
+#   void test(int foo) {
+#     return (foo << 24);
+#   }
+#
+#  In this case, the signed shift would be alright unless it would "push out"
+#  bits to the left, with undefined runtime result. Using ubsan, this can be
+#  caught.
```

Comment:
Are you sure this is undefined behavior? To my knowledge, this is perfectly fine. What would be undefined is

```C
void test(int foo) {
    return 1 << foo;
}
```

if `foo >= (sizeof(int) * 8) || foo < 0`.

Reply:
IIRC left-shifting int is undefined if the shifted int is negative:

6

In C bitwise left shift operation invokes Undefined Behaviour when the left side operand has negative value. and some.

ISO C99 (6.5.7/4):

    "The result of E1 << E2 is E1 left-shifted E2 bit positions; vacated bits are ﬁlled with zeros. If E1 has an unsigned type, the value of the result is E1 × 2E2, reduced modulo one more than the maximum value representable in the result type. If E1 has a signed type and nonnegative value, and E1 × 2E2 is representable in the result type, then that is the resulting value; otherwise, the behavior is undeﬁned.
"

Reply:
Strange. I was 100% sure that shifting with signed types is implementation defined regarding whether arithmetic or bitwise shifts are used, but not undefined.

The C++ guys did a good job with defining that signed integers are two's complement. There no relevant exceptions to that on the market anymore, and once the C standard would ban other representations the irrelevant exceptions would also cease.